### PR TITLE
Internal Java Solution for Loan Repayment

### PR DESCRIPTION
**Pull request by brad182**

### CODE REVIEW

**Review by dongliu0426 — APPROVED** [review]

File: solutions/usaco-991.mdx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+id: usaco-991
+source: USACO Silver 2020 January
+title: Loan Repayment
+author: Brad Ma
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_loan_silver_jan20.html)
+
+## Implementation
+**Time Complexity:** $\mathcal{O}(\sqrt{N}*logN)$
```
Comment (maggie-j-liu) [inline solutions/usaco-991.mdx:11]:
```suggestion
**Time Complexity:** $\mathcal{O}(\sqrt{N} \cdot \log{N})$
```

File: solutions/usaco-991.mdx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	static boolean isOk (long numGallons, long withinDays, long atLeast, long x) {
+		long g = 0;
+		while (withinDays > 0 && g < numGallons) {
+			long y = (numGallons - g) / x;
+			if (y < atLeast) {
+				long leftover = ((numGallons - g) + (atLeast - 1)) / atLeast;
+				return (leftover <= withinDays);
+			}
+			long maxMatch = (numGallons - (x * y));
+			long numDays = ((maxMatch - g) / (y)) + 1;
```
Comment (maggie-j-liu) [inline solutions/usaco-991.mdx:32]:
```suggestion
			long maxMatch = numGallons - (x * y);
			long numDays = (maxMatch - g) / y + 1;
```

File: solutions/usaco-991.mdx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+public class LoanRepayment {
+	// our checking function in the binary search
+	static boolean isOk (long numGallons, long withinDays, long atLeast, long x) {
+		long g = 0;
+		while (withinDays > 0 && g < numGallons) {
+			long y = (numGallons - g) / x;
+			if (y < atLeast) {
+				long leftover = ((numGallons - g) + (atLeast - 1)) / atLeast;
+				return (leftover <= withinDays);
```
Comment (maggie-j-liu) [inline solutions/usaco-991.mdx:29]:
```suggestion
				return leftover <= withinDays;
```

File: solutions/usaco-991.mdx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+			}
+			long maxMatch = (numGallons - (x * y));
+			long numDays = ((maxMatch - g) / (y)) + 1;
+			if (numDays > withinDays) {  // take a maximum
+				numDays = withinDays;
+			}
+			g += y * numDays;  // update values
+			withinDays -= numDays;
+		}
+		return (g >= numGallons);
```
Comment (maggie-j-liu) [inline solutions/usaco-991.mdx:39]:
```suggestion
		return g >= numGallons;
```

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Review by maggie-j-liu — APPROVED** [review]

File: solutions/usaco-991.mdx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class LoanRepayment {
+	// our checking function in the binary search
+	static boolean isOk (long numGallons, long withinDays, long atLeast, long x) {
```
Comment (SansPapyrus683) [inline solutions/usaco-991.mdx:23]: what's x now?

File: solutions/usaco-991.mdx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+<LanguageSection>
+  
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class LoanRepayment {
```
Comment (SansPapyrus683) [inline solutions/usaco-991.mdx:21]:
```suggestion
public class Loan {
```

File: solutions/usaco-991.mdx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio("loan");
+		long numGallons = io.nextLong();
+		long withinDays = io.nextLong();
+		long atLeast = io.nextLong();
+
+		long low = 1;
+		long high = 1000000000000L;
```
Comment (SansPapyrus683) [inline solutions/usaco-991.mdx:49]: i'd do integer.max_value but sure

File: solutions/usaco-991.mdx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+		// keep running until we narrowed it down to one number
```
Comment (SansPapyrus683) [inline solutions/usaco-991.mdx:51]: don't think we have to explain binary search here?

File: solutions/usaco-991.mdx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+		while (low < high) {
+			long mid = (low + high + 1) / 2;
+			// we can exclude everything up to but not including mid
+			if (isOk(numGallons, withinDays, atLeast, mid)) {
+				low = mid;
+			}
+			else {  // we can exclude everything including mid
```
Comment (SansPapyrus683) [inline solutions/usaco-991.mdx:58]:
```suggestion
			} else {  // we can exclude everything including mid
```

File: solutions/usaco-991.mdx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+```java
+import java.util.*;
+import java.io.*;
+
+public class LoanRepayment {
+	// our checking function in the binary search
+	static boolean isOk (long numGallons, long withinDays, long atLeast, long x) {
+		long g = 0;
+		while (withinDays > 0 && g < numGallons) {
+			long y = (numGallons - g) / x;
```
Comment (SansPapyrus683) [inline solutions/usaco-991.mdx:26]: explain some of the math here?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-991.mdx (hunk 7 above)
Comment (brad182) [reply solutions/usaco-991.mdx:49]: you need 10^12, not integer max value for this problem

File: solutions/usaco-991.mdx (hunk 9 above)
Comment (brad182) [reply solutions/usaco-991.mdx:58]: i dont do that format

File: solutions/usaco-991.mdx (hunk 10 above)
Comment (brad182) [reply solutions/usaco-991.mdx]: it is in the problem statement though

" 1. Assuming that Farmer John has already given Bessie G gallons, compute N−GX rounded down. Call this number Y.

   2. If Y is less than M, set Y to M.
   
   3. Give Bessie Y gallons of milk. "

Comment (SansPapyrus683) [reply solutions/usaco-991.mdx]: yeah that's kinda bad, just put some simple explanation
even just "compute amount of milk to give" would be nice

File: solutions/usaco-991.mdx (hunk 9 above)
Comment (SansPapyrus683) [reply solutions/usaco-991.mdx:57]: image

Comment (maggie-j-liu) [reply solutions/usaco-991.mdx:57]: nahh... i think it's fine as long as it's consistent...

Comment (SansPapyrus683) [reply solutions/usaco-991.mdx:57]: i mean nathan and michael said no so...

Comment (maggie-j-liu) [reply solutions/usaco-991.mdx:57]: i think they were just joking... right? 👀 @thecodingwizard @caoash

Comment (maggie-j-liu) [reply solutions/usaco-991.mdx:57]: image

Comment (SansPapyrus683) [reply solutions/usaco-991.mdx:57]: still looks pretty ugly imo

Comment (maggie-j-liu) [reply solutions/usaco-991.mdx:57]: image

Comment (SansPapyrus683) [reply solutions/usaco-991.mdx:57]: what's wrong w/ enforcing it though

Comment (HarryWangATX) [reply solutions/usaco-991.mdx:57]: nothing is, but it doesn't matter so the author can choose.

Comment (SansPapyrus683) [reply solutions/usaco-991.mdx:57]: ok lmao

**Review by nwatx — APPROVED** [review]
code looks ok

**Review by ryanchou-dev — APPROVED** [review]
formatting is good 👍

**Review by HarryWangATX — APPROVED** [review]

**Pull request merged**